File: packages/guards/src/retry-abuse.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface RetryAbuseOptions {
  action: 'block' | 'warn';
  maxRetries?: number;
  windowMs?: number;
}

export function retryAbuse(options: RetryAbuseOptions): Guard {
  const maxRetries = options.maxRetries ?? 5;
  const windowMs = options.windowMs ?? 10000;
  const requestLog: Map<string, number[]> = new Map();

  return {
    name: 'retry-abuse',
    version: '0.1.0',
    description: 'Detects retry abuse patterns including rapid retries and backoff violations',
    category: 'security',
    supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const now = Date.now();
      const key = text.trim().substring(0, 200);
      const issues: string[] = [];

      const timestamps = requestLog.get(key) ?? [];
      timestamps.push(now);

      const recent = timestamps.filter((t) => now - t <= windowMs);
      requestLog.set(key, recent);

      if (recent.length > maxRetries) {
        issues.push('max_retries_exceeded');
      }

      if (recent.length >= 3) {
        const gaps: number[] = [];
        for (let i = 1; i < recent.length; i++) {
          gaps.push(recent[i] - recent[i - 1]);
        }
        const allRapid = gaps.every((g) => g < 100);
        if (allRapid) {
          issues.push('identical_rapid_retries');
        }

        if (gaps.length >= 3) {
          const increasing = gaps.every((g, i) => i === 0 || g >= gaps[i - 1]);
          if (!increasing && recent.length > 3) {
            issues.push('backoff_violation');
          }
        }
      }

      const triggered = issues.length > 0;
      const score = triggered ? Math.min(issues.length / 2, 1.0) : 0;

      return {
        guardName: 'retry-abuse',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues, retryCount: recent.length } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/retry_abuse.py
"""Detect retry abuse patterns including rapid retries."""
from __future__ import annotations

import time
from typing import Optional

from open_guardrail.core import GuardResult


class _RetryAbuse:
    def __init__(
        self,
        *,
        action: str = "block",
        max_retries: int = 5,
        window_ms: int = 10000,
    ) -> None:
        self.name = "retry-abuse"
        self.action = action
        self._max_retries = max_retries
        self._window_ms = window_ms
        self._log: dict[str, list[float]] = {}

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        now = time.time() * 1000
        key = text.strip()[:200]
        issues: list[str] = []

        timestamps = self._log.get(key, [])
        timestamps.append(now)
        cutoff = now - self._window_ms
        recent = [t for t in timestamps if t >= cutoff]
        self._log[key] = recent

        if len(recent) > self._max_retries:
            issues.append("max_retries_exceeded")

        if len(recent) >= 3:
            gaps = [
                recent[i] - recent[i - 1]
                for i in range(1, len(recent))
            ]
            if all(g < 100 for g in gaps):
                issues.append("identical_rapid_retries")
            if len(gaps) >= 3:
                increasing = all(
                    gaps[i] >= gaps[i - 1]
                    for i in range(1, len(gaps))
                )
                if not increasing and len(recent) > 3:
                    issues.append("backoff_violation")

        triggered = len(issues) > 0
        score = min(len(issues) / 2, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="retry-abuse",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Retry abuse detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "issues": issues,
                    "retry_count": len(recent),
                }
                if triggered
                else None
            ),
        )


def retry_abuse(
    *,
    action: str = "block",
    max_retries: int = 5,
    window_ms: int = 10000,
) -> _RetryAbuse:
    return _RetryAbuse(
        action=action,
        max_retries=max_retries,
        window_ms=window_ms,
    )
